src/pmdas/systemtap/pmdasystemtap.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2008 Aconex.  All Rights Reserved.
.\" 
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\" 
.\"
.TH PMDASYSTEMTAP 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmdasystemtap\f1 \- systemtap performance metrics domain agent (PMDA)
.SH DESCRIPTION
\f3pmdasystemtap\f1 is a Performance Metrics Domain Agent (\s-1PMDA\s0)
which exports metric values from the Linux systemtap dynamic tracing toolkit.
.PP
This implementation uses the
.BR stap (1)
tool, which is a front-end to the systemtap toolkit.
.SH INSTALLATION
In order to access performance data exported by systemtap from
with PCP, it is necessary to perform two configuration steps.
.PP
Firstly, configure systemtap probes, and verify them with
.BR stap (1) .
These should be produced in a format that is easily parsed,
and then stored in the
.I $PCP_PMDAS_DIR/systemtap/probes.stp
file.
.PP
Then
.B pmdasystemtap
must be configured to extract the values from the text produced by
.BR stap .
Two example probes are implemented in the default systemtap PMDA
script - readdir and sync traces (see
.I $PCP_PMDAS_DIR/systemtap/pmdasystemtap.pl
for details).
.PP
      # cd $PCP_PMDAS_DIR/systemtap
.br
      # [ edit probes.stp, test /usr/bin/stap probes.stp ]
.br
      # [ edit pmdasystemtap.pl ]
.PP
Once this is setup, you can access the names and values for the
systemtap performance metrics by doing the following as root:
.PP
      # cd $PCP_PMDAS_DIR/systemtap
.br
      # ./Install
.PP
To uninstall, do the following as root:
.PP
      # cd $PCP_PMDAS_DIR/systemtap
.br
      # ./Remove
.PP
\fBpmdasystemtap\fR is launched by \fIpmcd\fR(1) and should never be executed 
directly. The Install and Remove scripts notify \fIpmcd\fR(1) when the 
agent is installed or removed.
.SH FILES
.IP "\fB$PCP_PMDAS_DIR/systemtap/probes.stp\fR" 4 
probe configuration file for \fBstap\fR(1), run by \fBpmdasystemtap\fR
.IP "\fB$PCP_PMDAS_DIR/systemtap/Install\fR" 4 
installation script for the \fBpmdasystemtap\fR agent 
.IP "\fB$PCP_PMDAS_DIR/systemtap/Remove\fR" 4 
undo installation script for the \fBpmdasystemtap\fR agent 
.IP "\fB$PCP_LOG_DIR/pmcd/systemtap.log\fR" 4 
default log file for error messages from \fBpmdasystemtap\fR 
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fR are used to parameterize
the file and directory names used by \fBPCP\fR. On each installation, the
file \fB/etc/pcp.conf\fR contains the local values for these variables. 
The \fB$PCP_CONF\fR variable may be used to specify an alternative 
configuration file, as described in \fIpcp.conf\fR(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmcd (1)
and
.BR stap (1).


CHANGELOG
pcp-3.10.0 (15 October 2014)
    - Currently under development.

pcp-3.9.10 (5 September 2014)
    - Support for native systemd service files for daemons
    - Support for nanosecond resolution event timestamps
    - Changes to the FreeBSD build to bring it up to date
    - libpcp: reduce daemon memory usage with secure sockets
    - libpcp: add auto-deflate support for xz compression
    - pmie/pmlogger scripts: xz compression support enabled
    - pmie/pmlogger scripts: lzo compression support added
    - pmlogconf: fix slow pmlogger_daily remote host startup
    - pmcd_wait: fixed to handle af_unix sockets correctly
    - pmatop: fix help keys
    - pmatop: parse standard pcp options (including archives)
    - pmstore: support for local-context mode of operation
    - pmdiff: additional reporting options, added to /usr/bin
    - pmdiff: fix column layout with custom precision setting
    - pmlogsummary: fix time window reporting with -S and -T
    - pmdapapi: new agent for PAPI hardware event counters
    - pmdagfs2: latency metrics for glock grant/demote/queue
    - pmdamysql: change metric semantics for "uptime"
    - pmdaxfs: fix DSO mode and allow test data injection
    - pmdammv: long option support
    - docs: numerous updates to the online documentation
    - docs: added a quick reference guide for Red Hat users

pcp-3.9.9 (13 August 2014)
    - sar2pcp: support for systat version 10.2.0
    - pmiostat(1): new monitor tool with iostat-alike output
    - pcp-dmcache(1): new monitor tool for dm-cache devices
    - pmdalinux: disk.dm.* metric hierarchiy (device mapper)
    - pmdalinux: refine device mapper naming for old platforms
    - pmdalinux: allow controlled data injection for QA tests
    - pmdadmcache: new agent for device mapper cache targets
    - pmdainfiniband: correctly initialise switch metrics
    - libpcp: global timeout option for service discovery
    - libpcp: __pmDiscoverServicesWithOptions() improvements
    - libpcp: grok '%<interface' for link-local ipv6 addresses
    - libpcp: derived metrics support for new rate() keyword
    - python api: many convenience class module extensions
    - python api: timeval accessor helper interfaces
    - python api: pmUnitStr now uses thread-safe version
    - python api: pmNumberStr_r now available from python
    - python api: fix derived metrics interface exceptions
    - pmlogconf: mark some log-default metrics as log-once
    - dbpmda: use the correct (blocking) socket API variants
    - pmlogger_daily: fix $PCP_RUN_DIR creation checks

pcp-3.9.8 (16 July 2014)
    - Add overlooked Debian packaging metadata updates.

pcp-3.9.7 (16 July 2014)
    - pmie: "ruleset" extensions to the pmie(1) language
    - pmdaproc: additional memory and new blkio cgroup metrics
    - pmdanvidia: new Nvidia GPU PMDA using NVML interfaces
    - pmdainfiniband: add switch counter metrics
    - pmdalustrecomm: bug fixes in the lnet memory metrics
    - pmdamemcache: fixed underlying socket code
    - pmdalinux: mem.util.available for down-rev kernels
    - pmdagfs2: several fixes and improvements in behaviour
    - pmdas: support for long form command line options
      (aix, jbd2, linux, lustre, mounts, processes, sample,
      sendmail, txmon, xfs)
    - libpcp: 'timeout' option for 'probe' service discovery
    - libpcp: secure sockets timeout handling improvements
    - pmie examples: updates to the pmie tutorial docs
    - pmchart: new MemAvailable view for mem.util.available
    - porting: additional GNU/Hurd port updates
    - packaging: fix Makepkgs script for old tar versions
    - packaging: fix rpm builds for rpm versions prior to 4.6
    - libpcp: use minimum required stack for discovery threads
    - libpcp: archive folio parsing in option handling code
    - python APIs: archive folio parsing for recording tools
    - man pages: updates to service discovery APIs and tools
    - libpcp: archive interpolation thread safety improvements
    - libpcp: fixed memory leak from pmDestroyContext(3)

pcp-3.9.6 (19 June 2014)
    - Fix pthreads build fallout for Solaris platforms
    - Remove accidental /var/run/pcp install on Debian
    - Rename configure.in to configure.ac as autoconf prefers

pcp-3.9.5 (18 June 2014)
    - porting work on the GNU/Hurd platform
    - porting work on Windows (early MinGW-64 work)
    - pmdalinux: add in the new mem.util.available metric
    - pmchart: new vCPU view, resolve guest time reporting
    - pmwebd: add CORS "Access-Control-Allow-Origin: *" to
      JSON response headers
    - pmatop: improvements to value fitting
    - packaging: gui debuginfo/testsuite correctly replaced
    - pmconfig: moved install location for Solaris conflict
    - pmstat: fix hostname reporting when in cluster mode
    - Coverity scan fixes in libpcp, python, and elsewhere
    - packaging: pcp-doc now contains all of the tutorial bits
    - python: timeval/metricspec error handling improvements
    - pmatop: handle curses problems more gracefully
    - pmdapostfix: run as postfix user, not the pcp user
    - python: fix pmParseInterval failure segfault
    - python: fix segfault when printing empty units
    - libpcp: functions for collecting server port numbers
    - pmcd.services metric reflecting PCP daemon usage
    - libpcp, pmfind: active service probing
    - pmproxy, pmwebd: support for service discovery
    - pmcollectl, pmatop: improved log writing support
    - pmie: long option support
    - pmquery: neater output formatting
    - pmchart, pmtime: long option support
    - pmdasystemd: install the built binary
    - python: significant work on the pcp.pmcc module, adding
      long option support, scale conversion, an initial fetch
      loop method, and entirely new convenience classes
    - pmlogcheck: fix bad memory access sigsegv on some logs
    - Fix infinite loop in build scripts without Qt enabled
    - pmdalinux: add metrics from /proc/net/netstat
    - pmdumptext: long option support

src/pmdas/systemtap/pmdasystemtap.pl
#
# Copyright (c) 2012 Red Hat.
# Copyright (c) 2008 Aconex.  All Rights Reserved.
# 
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 

use strict;
use warnings;
use PCP::PMDA;

use vars qw( $pmda $id );
my $probe_indom = 0;
my $probe_script = pmda_config('PCP_PMDAS_DIR') . '/systemtap/probes.stp';
my $probe_command = "/usr/bin/stap -m pmdasystemtap $probe_script";
my @probe_instances = ( 0 => 'sync', 1 => 'readdir' );
my ( $sync_count, $sync_pid, $sync_cmd ) = ( 0, 0, "(none)" );
my ( $readdir_count, $readdir_pid, $readdir_cmd ) = ( 0, 0, "(none)" );

sub systemtap_input_callback
{
    ( $id, $_ ) = @_;
    # $pmda->log($_);

    if (/^readdir: \((\d+)\) (.*)$/) {
	( $readdir_pid, $readdir_cmd ) = ( $1, $2 );
	$readdir_count++;
    }
    elsif (/^sync: \((\d+)\) (.*)$/) {
	( $sync_pid, $sync_cmd ) = ( $1, $2 );
	$sync_count++;
    }
}

sub systemtap_fetch_callback
{
    my ($cluster, $item, $inst) = @_;

    if ($inst < 0 || $inst > 1)	{ return (PM_ERR_INST, 0); }
    if ($cluster == 0) {
	if ($item == 0)	{
	    if ($inst == 0)	{ return ($sync_count, 1); }
	    else		{ return ($readdir_count, 1); }
	}
	elsif ($item == 1) {
	    if ($inst == 0)	{ return ($sync_pid, 1); }
	    else		{ return ($readdir_pid, 1); }
	}
	elsif ($item == 2) {
	    if ($inst == 0)	{ return ($sync_cmd, 1); }
	    else		{ return ($readdir_cmd, 1); }
	}
    }
    return (PM_ERR_PMID, 0);
}

$pmda = PCP::PMDA->new('systemtap', 88);

$pmda->add_metric(pmda_pmid(0,0), PM_TYPE_U32, $probe_indom,
		  PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
		  'systemtap.probes.count',
		  'Number of times the probe has been observed', '');
$pmda->add_metric(pmda_pmid(0,1), PM_TYPE_32, $probe_indom,
		  PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
		  'systemtap.probes.pid',
		  'The PID of the last process to pass the probe point', '');
$pmda->add_metric(pmda_pmid(0,2), PM_TYPE_STRING, $probe_indom,
		  PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
		  'systemtap.probes.cmd',
		  'The name of the last process to pass the probe point', '');

$pmda->add_indom($probe_indom, \@probe_instances,
		 'Instance domain exporting each SystemTap probe', '');

$pmda->set_fetch_callback(\&systemtap_fetch_callback);
$pmda->add_pipe($probe_command, \&systemtap_input_callback, 0);
$pmda->set_user('pcp');
$pmda->run;
